src/Init/Data/BitVec/Bitblast.lean

```diff
@@ -326,4 +326,114 @@ theorem getLsb_mul (x y : BitVec w) (i : Nat) :
   · simp
   · omega
 
+/-## shiftLeft recurrence for bitblasting -/
+
+/--
+A recurrence that describes shifting by an arbitrary bitvector
+as shifting by a constant amount.
+
+This computation is equivalent to `x <<< (y.truncate (n + 1)).zeroExtend w₂`,
+as witnessed by the `shiftLeftRec_eq` theorem further down.
+We phrase it recursively to simplify other proofs
+ -/
+def shiftLeftRec (x : BitVec w₁) (y : BitVec w₂) (n : Nat) : BitVec w₁ :=
+  let shiftAmt := (y &&& (twoPow w₂ n))
+  match n with
+  | 0 => x <<< shiftAmt
+  | n + 1 => (shiftLeftRec x y n) <<< shiftAmt
+
+@[simp]
+theorem shiftLeftRec_zero {x : BitVec w₁} {y : BitVec w₂} :
+    shiftLeftRec x y 0 = x <<< (y &&& twoPow w₂ 0)  := by
+  simp [shiftLeftRec]
+
+@[simp]
+theorem shiftLeftRec_succ {x : BitVec w₁} {y : BitVec w₂} :
+    shiftLeftRec x y (n + 1) =
+      (shiftLeftRec x y n) <<< (y &&& twoPow w₂ (n + 1)) := by
+  simp [shiftLeftRec]
+
+@[simp]
+theorem getLsb_ofNat_one {w i : Nat} :
+    (1#w).getLsb i = (decide (i = 0) && decide (i < w)) := by
+  rcases w with rfl | w
+  · simp
+  · simp only [getLsb, toNat_ofNat, testBit_mod_two_pow]
+    by_cases hi : i = 0
+    · simp [hi]
+    · simp only  [hi, decide_False, Bool.false_and,
+        and_eq_false_imp, decide_eq_true_eq]
+      intros _
+      simp only [testBit, shiftRight_eq_div_pow, one_and_eq_mod_two, bne_eq_false_iff_eq]
+      suffices 1 / 2^i = 0 by simp [this]
+      apply Nat.div_eq_of_lt (Nat.one_lt_two_pow_iff.mpr hi)
+
+/--
+If `(y &&& z = 0)`, then shifting by `y ||| z` is the same as shifting by `y` and then by `z`.
+Note that the hypothesis `h'` is implied by `h : y &&& z = 0#w`,
+but we choose to take the additional hypothesis and leave proving it
+as an implication of `h` for a follow up PR.
```

Comment:
This is an easy consequence of `add_eq_or_of_and_eq_zero`, no? It's just a line or two, easier to write it than to write this doc-string!

Reply:
Particularly given that below you write a 20 line proof of this in a special case??

Reply:
Oops, I see, it's not implied by that.